### PR TITLE
ACS CTE: Add forward modeling functionality

### PR DESCRIPTION
**Pull request by jamienoss**
The cmd line option ``--forwardModelOnly``
piggy-backs upon the normal CTE code and usage (gen2 only) however,
instead of correcting the CTE trails it actually simulates (adds) them.

All CTE flags and ref files must be defined in the exact same
manner needed to run the actual CTE correction.

``PCTECORR`` must be set to ``PERFORM``.

Signed-off-by: James Noss <jnoss@stsci.edu>

### CODE REVIEW

**Comment by jamienoss** [comment]
@mdlpstsci & @pllim This is still a little hacky and there's duplicated code that _could_ be factored, but anyhow here it is.

**Comment by jamienoss** [comment]
I decided to revert back to requiring ``PCTECORR=PERFORM`` on the basis that if no action flags are set to ``PERFORM``, hstcal complains (at some early and generic stage) and exits. So if all ``XXXCORR`` equal either ``OMIT`` or, more relevant ``COMPLETE``, something needs to be set to ``PERFORM`` and this _should_ be ``PCTECORR``.

**Comment by jamienoss** [comment]
An error may need to be added for when using ``--forwardModelOnly`` and the gen 1 algorithm. If not, gen1 use will ignore the flag do the actual CTE without the user knowing. Actually, I've just thought about where this could be added, so will quickly do now.

File: pkg/acs/calacs/calacs/acsmain.c [hunk 1]
```diff
@@ -25,7 +26,7 @@ char MsgText[MSG_BUFF_LENGTH]; // Global char auto initialized to '\0'
 
 static void printSyntax(void)
 {
-    printf("syntax:  calacs.e [--help] [-t] [-s] [-v] [-q] [-r] [--version] [--gitinfo] [-1|--nthreads <N>] [--ctegen <1|2>] [--pctetab <path>] input \n");
+    printf("syntax:  calacs.e [--help] [-t] [-s] [-v] [-q] [-r] [--version] [--gitinfo] [-1|--nthreads <N>] [--ctegen <1|2>] [--pctetab <path>] [--forwardModelOnly] input \n");
```
Comment (pllim) [inline pkg/acs/calacs/calacs/acsmain.c]: What is the use case of running the entire pipeline on RAW using forward model? I don't think this needs to be an option for `calacs.e`?

Comment (jamienoss) [reply pkg/acs/calacs/calacs/acsmain.c]: Good point, I'll remove.

**Comment by pllim** [comment]
> there's duplicated code that could be factored

Would be nice to address the code duplication sooner than later, IMHO.

**Comment by jamienoss** [comment]
> I know how you and Michele feels about those Python wrappers but to be consistent, this might also need a wrapper over at ACSTOOLS.

I'll wrap away.

> I was thinking more like a new, say, acsctefowardmodel.e command, as not to confuse users too much. acscte.e turns blv_tmp into blc_tmp

I'll create a wrapper for this too.

>turns blv_tmp into blc_tmp.

This still creates a blc_tmp. which would need to be addressed before merge. Not sure what the team will want.

**Comment by mdlpstsci** [comment]
I will be able to do a full review next week.

To clarify: I am not against Python wrappers per se, I am against duplication of code.

I would agree with @pllim that this is not really a calacs function.  It should be a standalone routine as it is for modelling purposes.  I can hopefully say more intelligent things once I review the code.

I would say this overall project was the biggest success story of the Hack Day! Kudos.

**Comment by jamienoss** [comment]
@pllim 
>>I was thinking more like a new, say, acsctefowardmodel.e command, as not to confuse users too much. acscte.e turns blv_tmp into blc_tmp
>
>I'll create a wrapper for this too.

Actually, nope, maybe I'll let someone else do that.

**Comment by jamienoss** [comment]
>>there's duplicated code that could be factored
>
>Would be nice to address the code duplication sooner than later, IMHO.

Ok, I lied about the _could_. The code isn't easy to dedup (if at all) and prob. isn't worth it or even trying. The main argument is the diff in memory allocations.

File: pkg/acs/calacs/acscte/dopcte-gen2.c [hunk 2]
```diff
@@ -201,13 +221,26 @@ int doPCTEGen2 (ACSInfo *acs, CTEParamsFast * ctePars, SingleGroup * chipImage)
...
-       trlmessage("(pctecorr) Running correction algorithm...");
-       //perform CTE correction
        SingleGroup * cteCorrectedImage = &columnMajorImage;
-       if ((status = inverseCTEBlur(&smoothedImage, cteCorrectedImage, &trapPixelMap, ctePars)))
+       if (forwardModelOnly)
        {
-           freeOnExit(&ptrReg);
-           return status;
+           trlmessage("(pctecorr) Running forward model simulation...");
+           //perform CTE correction
```
Comment (mdlpstsci) [inline pkg/acs/calacs/acscte/dopcte-gen2.c]: Minor issue: Comment should be updated.

**Comment by jamienoss** [comment]
I had forgotten to ``git add`` the new ``wscript``. @mdlpstsci I wouldn't bother re-reviewing it till we hear from the team about the file name.

File: pkg/acs/calacs/acscte/forward_model/wscript [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+                  '../dopcte.c',
+                  '../dopcte-gen2.c',
+                  '../getcteflag.c',
+                  '../getctesw.c',
+                  '../pcte_fixycte.c',
+                  '../pcte_funcs.c',
+                  '../maincte.c',
+                  'main.c'
+                  ],
+        target = 'acscteforwardmodel.e',
```
Comment (jamienoss) [inline pkg/acs/calacs/acscte/forward_model/wscript]: @mdlpstsci This is where the new exec is named.

**Comment by jamienoss** [comment]
The ACS team decided on ``_ctefmod.fits``.

File: pkg/acs/calacs/calacs/calacs.c [hunk 4]
```diff
@@ -611,7 +612,7 @@ int ProcessACSCCD (AsnInfo *asn, ACSInfo *acshdr, int *save_tmp, int printtime,
                 if (acshdr->sci_basic_cte == PERFORM) {
                     if (ACScte(acshdr->blv_tmp, acshdr->blc_tmp,
                                &acscte_sci_sw, &sciref, printtime,
-                               asn->verbose, nThreads, cteAlgorithmGen, pcteTabNameFromCmd)) { //this is the line
+                               asn->verbose, nThreads, cteAlgorithmGen, pcteTabNameFromCmd, false)) { //this is the line
```
Comment (mdlpstsci) [inline pkg/acs/calacs/calacs/calacs.c]: I would like to see an explicit comment here indicating the forwardModelOnly is not a feature that was designed to be used during pipeline processing (hence the false).  Rather this option is for standalone processing when running the CTE component of the pipeline for calibration purposes.

**Comment by jamienoss** [comment]
@mdlpstsci back to you for final review.

**Pull request merged**